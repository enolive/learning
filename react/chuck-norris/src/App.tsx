import React, {useEffect, useReducer} from 'react';
import logo from './logo.jpg';
import './App.css';

export interface Joke {
  value: {
    joke: string
  }
}

export const tidy = (text: string) =>
  text.replace(/&quot;/g, '"');

type State =
  | { status: 'empty' }
  | { status: 'loading' }
  | { status: 'success', jokeText: string }
  | { status: 'failure', error: string }

type Action =
  | { type: 'fetch' }
  | { type: 'success', result: string }
  | { type: 'failure', error: string };

const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case 'success':
      return {status: 'success', jokeText: action.result};
    case 'failure':
      return {status: 'failure', error: action.error};
    case 'fetch':
      return {status: 'loading'};
  }
};

function App() {
  const [jokeData, dispatch] = useReducer(reducer, {status: 'empty'});
  const fetchRandomJoke = async () => {
    dispatch({type: 'fetch'});
    await fetch('http://api.icndb.com/jokes/random')
      .then(response => response.json() as Promise<Joke>)
      .then(joke => joke.value.joke)
      .then(tidy)
      .then(joke => dispatch({type: 'success', result: joke}))
      .catch((e: Error) => dispatch({type: 'failure', error: e.message}));
  };

  useEffect(() => {
    fetchRandomJoke();
  }, []);

  const refreshJoke = async () => {
    await fetchRandomJoke();
  };

  return (
    <div className="App">
      <div className="container-fluid">
        <header className="row">
          <figure className="figure">
            <img className="figure-caption" src={logo} alt="Chuck Norris Logo"/>
            <figcaption className="figure-caption">
              <a data-testid="attribution"
                 href="https://www.flickr.com/photos/enriquespics/2058486062">
                by Carlos Killpack
              </a>
            </figcaption>
          </figure>
        </header>
        <main className="row">
          {jokeData.status === 'success' && <p data-testid="joke-text">Random Wisdom: {jokeData.jokeText}.</p>}
          {jokeData.status === 'failure' && <p data-testid="error">{jokeData.error}</p>}
        </main>
        <footer className="row">
          <button className="btn btn-primary"
                  onClick={refreshJoke}
                  disabled={jokeData.status === 'loading'}>
            Refresh
          </button>
        </footer>
      </div>
    </div>
  );
}

export default App;
