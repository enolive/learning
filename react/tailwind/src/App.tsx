import React, {FC} from 'react'

const App: FC = () => (
    <div className="h-screen bg-amber-600">
        <h1 className="m-6 text-amber-50 text-5xl bg-amber-800 rounded">Welcome to my App!</h1>
        <p className="font-extrabold">
        just a simple test for tailwind
        </p>
    </div>
)

export default App
