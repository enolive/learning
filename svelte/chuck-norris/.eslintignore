rollup.config.js
svelte.config.js
.eslintrc.js
