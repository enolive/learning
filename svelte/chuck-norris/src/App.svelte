<script lang="ts">
  interface JokeText {
    value: string;
  }

  function fetchRandomJoke() {
    fetchJoke$ = fetch('https://api.chucknorris.io/jokes/random')
      .then(res => res.json() as JokeText)
      .then(joke => joke.value);
  }

  let fetchJoke$: Promise<string>;

  fetchRandomJoke();
</script>

<header>
  <h1>Chuck Norris</h1>
</header>
<main>
  <figure>
    <img src="chuck-norris-logo.png" alt="Chuck Norris approves!">
    <figcaption>Chuck Norris approves!</figcaption>
  </figure>
  <p role="status">
    {#await fetchJoke$}
      Loading...
    {:then jokeText}
      Random Wisdom: {jokeText}
    {/await}
  </p>
  <button on:click={fetchRandomJoke}>Refresh</button>
</main>
