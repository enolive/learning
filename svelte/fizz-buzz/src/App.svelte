<script lang="ts">
  import {fizzBuzz} from './fizzbuzz';

  let input: number;
  let result: string;

  const calculateIt = () => {
    result = fizzBuzz(input);
  };
</script>

<main>
    <h1>Fizz Buzz</h1>
    <form on:submit|preventDefault={calculateIt}>
        <label>
            Enter a number:
            <input type="number" bind:value={input}>
        </label>
        <button>Calculate It</button>
        {#if result}
          <div role="status">Result: {result}</div>
        {/if}
    </form>
</main>

<style>
    main {
        text-align: center;
        padding: 1em;
        max-width: 240px;
        margin: 0 auto;
    }

    h1 {
        color: #ff3e00;
        text-transform: uppercase;
        font-size: 4em;
        font-weight: 100;
    }

    @media (min-width: 640px) {
        main {
            max-width: none;
        }
    }
</style>
