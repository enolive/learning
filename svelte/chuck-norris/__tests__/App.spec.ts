import App from '../src/App.svelte';
import {fireEvent, render, waitFor} from '@testing-library/svelte';

beforeEach(() => {
  fetchMock.resetMocks();
})

test('renders correctly', () => {
  const {getByRole} = render(App, {});

  expect(getByRole('heading')).toHaveTextContent('Chuck Norris');
  expect(getByRole('figure')).toMatchSnapshot();
  expect(getByRole('status')).toHaveTextContent('Loading...');
  expect(getByRole('button')).toHaveTextContent('Refresh');
});

test('fetches jokes by api', () => {
  const jokeText = 'There\'s no place like 127.0.0.1...';
  fetchMock.mockResponseOnce(JSON.stringify({value: jokeText}));

  const {getByRole} = render(App);

  expect(fetchMock).toHaveBeenCalledWith('https://api.chucknorris.io/jokes/random');
  waitFor(() => {
    expect(getByRole('status')).toHaveTextContent(`Random wisdom: ${jokeText}`);
  });
});

test('refreshes joke', () => {
  const {getByRole} = render(App);
  const refreshButton = getByRole('button');

  fireEvent.click(refreshButton);

  expect(fetchMock).toHaveBeenCalledTimes(2);
});

