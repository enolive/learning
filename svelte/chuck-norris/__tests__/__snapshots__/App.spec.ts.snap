// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<figure>
  <img
    alt="Chuck Norris approves!"
    src="chuck-norris-logo.png"
  />
   
  <figcaption>
    Chuck Norris approves!
  </figcaption>
</figure>
`;
